### ISSUE TITLE
Queue waiting for data: (0, points2)

### ISSUE THREAD

**Issue by nitramek** [googlecartographer/cartographer#722]
Hi,

My setup (right now) is single VLP16 + Android phone as IMU (to be replaced, now just for testing). I am trying to make cartographer work and then try to calibrate settings and such. Also running on Ubuntu 16.04 LTS, ROS Kinetic. In online mode I run velodyne driver and each UDP packet is send as PointCloud2 as suggested.

My goal is to get 3D SLAM up and running but right now it gives me this warning and cartographer is producing no output. I have tried to debug the pipeline with some print statements but with no sucess. I found out the data is coming through but thats it.

The exact configuration lies in forked repo in the nitramek branch.
[https://github.com/nitramek/cartographer_ros/tree/nitramek/cartographer_ros](https://github.com/nitramek/cartographer_ros/tree/nitramek/cartographer_ros)
Direct links:
- [Launch file](https://github.com/nitramek/cartographer_ros/blob/nitramek/cartographer_ros/launch/cart_test.launch)
- [URDF](https://github.com/nitramek/cartographer_ros/blob/nitramek/cartographer_ros/urdf/cart_test.urdf) - doesnt reflect the setup exactly the IMU is really close to the VLP sensor. Right now I am trying to just have some output. 
- [LUA config](https://github.com/nitramek/cartographer_ros/blob/nitramek/cartographer_ros/configuration_files/cart_test.lua) - I have read that if the data from the VLP is divided in each packet the num_accumulated_range_data should be higher but I tried that configuration too and it didnt help.

Output of roslaunch with cart_test.bag
```roslaunch cartographer_ros cart_test.launch bag_filename:=/home/nitramek/vsb/bags/cart_test.bag 
... logging to /home/nitramek/.ros/log/554263dc-d60e-11e7-8959-000c29996d44/roslaunch-ubuntu-22028.log
Checking log directory for disk usage. This may take awhile.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://ubuntu:41761/

SUMMARY
========

PARAMETERS
 * /robot_description: <...>
 * /rosdistro: kinetic
 * /rosversion: 1.12.7
 * /use_sim_time: True

NODES
  /
    cartographer_node (cartographer_ros/cartographer_node)
    cartographer_occupancy_grid_node (cartographer_ros/cartographer_occupancy_grid_node)
    playbag (rosbag/play)
    robot_state_publisher (robot_state_publisher/robot_state_publisher)
    rviz (rviz/rviz)

auto-starting new master
process[master]: started with pid [22039]
ROS_MASTER_URI=http://localhost:11311

setting /run_id to 554263dc-d60e-11e7-8959-000c29996d44
process[rosout-1]: started with pid [22052]
started core service [/rosout]
process[robot_state_publisher-2]: started with pid [22056]
process[cartographer_node-3]: started with pid [22061]
process[cartographer_occupancy_grid_node-4]: started with pid [22063]
process[rviz-5]: started with pid [22067]
process[playbag-6]: started with pid [22081]
[ INFO] [1512074271.288926372]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer_ros/configuration_files/cart_test.lua' for 'cart_test.lua'.
[ INFO] [1512074271.290025782]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/map_builder.lua' for 'map_builder.lua'.
[ INFO] [1512074271.290178792]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/map_builder.lua' for 'map_builder.lua'.
[ INFO] [1512074271.290376293]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/pose_graph.lua' for 'pose_graph.lua'.
[ INFO] [1512074271.290455295]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/pose_graph.lua' for 'pose_graph.lua'.
[ INFO] [1512074271.290619226]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/trajectory_builder.lua' for 'trajectory_builder.lua'.
[ INFO] [1512074271.290729493]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/trajectory_builder.lua' for 'trajectory_builder.lua'.
[ INFO] [1512074271.290861462]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/trajectory_builder_2d.lua' for 'trajectory_builder_2d.lua'.
[ INFO] [1512074271.290935842]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/trajectory_builder_2d.lua' for 'trajectory_builder_2d.lua'.
[ INFO] [1512074271.291261740]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/trajectory_builder_3d.lua' for 'trajectory_builder_3d.lua'.
[ INFO] [1512074271.291402077]: I1130 21:37:51.000000 22061 configuration_file_resolver.cc:41] Found '/home/nitramek/ros_all/cartographer/install_isolated/share/cartographer/configuration_files/trajectory_builder_3d.lua' for 'trajectory_builder_3d.lua'.
[ INFO] [1512074271.332676621]: I1130 21:37:51.000000 22061 submaps.cc:295] Added submap 1
[ INFO] [1512074271.332857002]: I1130 21:37:51.000000 22061 map_builder_bridge.cc:101] Added trajectory with ID '0'.
[ INFO] [1512074271.736876989, 1511821305.741212741]: I1130 21:37:51.000000 22061 ordered_multi_queue.cc:172] All sensor data for trajectory 0 is available starting at '636475895586160000'.
[ WARN] [1512074277.194455759, 1511821311.204946129]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.227861131, 1511821311.236593011]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.302636897, 1511821311.311198297]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.382513099, 1511821311.395140797]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.464577494, 1511821311.478894225]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.554863906, 1511821311.563483827]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.628020913, 1511821311.637643983]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.666354153, 1511821311.680797209]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.740038747, 1511821311.754346146]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.807312226, 1511821311.817188198]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.887312926, 1511821311.901671975]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074277.968685009, 1511821311.977444411]: W1130 21:37:57.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.046435695, 1511821312.052054020]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.125584790, 1511821312.135644610]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.151880098, 1511821312.156721889]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.225074175, 1511821312.230574158]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.302265204, 1511821312.313890867]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.382167219, 1511821312.389354254]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.469980845, 1511821312.484263222]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.540587075, 1511821312.548092741]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.599466833, 1511821312.613347019]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.646836600, 1511821312.655329673]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.722752032, 1511821312.730905215]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.802007193, 1511821312.815672655]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.881915609, 1511821312.890814483]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074278.945044962, 1511821312.954728026]: W1130 21:37:58.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.002306841, 1511821313.007231263]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.073386705, 1511821313.082975012]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.148572448, 1511821313.158120288]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.214656487, 1511821313.221299647]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.292562095, 1511821313.307078457]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.355294195, 1511821313.360917874]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.427769001, 1511821313.434860344]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.495473399, 1511821313.507317047]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.571443194, 1511821313.581611706]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.637106823, 1511821313.643824492]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.714402779, 1511821313.727557616]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.777421967, 1511821313.791736190]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.846749518, 1511821313.854807320]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.922302492, 1511821313.930062918]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074279.990129517, 1511821314.004590758]: W1130 21:37:59.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.058610109, 1511821314.069531202]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.129311475, 1511821314.134185939]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.200541830, 1511821314.207334552]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.279816610, 1511821314.293888091]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.347228074, 1511821314.358568039]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.420317376, 1511821314.432818963]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.495046546, 1511821314.508127398]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.585604628, 1511821314.591724952]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.640162309, 1511821314.654408148]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.708775422, 1511821314.717442272]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.786608088, 1511821314.792751875]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.852228104, 1511821314.866511071]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.921965420, 1511821314.929493985]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074280.992212597, 1511821315.002391782]: W1130 21:38:00.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074281.062716332, 1511821315.076861552]: W1130 21:38:01.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074281.133521486, 1511821315.138974962]: W1130 21:38:01.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074281.206554835, 1511821315.211032640]: W1130 21:38:01.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074281.272430631, 1511821315.284573778]: W1130 21:38:01.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074281.348050065, 1511821315.358553932]: W1130 21:38:01.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074281.420092099, 1511821315.432730734]: W1130 21:38:01.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074281.485823777, 1511821315.496987119]: W1130 21:38:01.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[ WARN] [1512074281.557036777, 1511821315.570777630]: W1130 21:38:01.000000 22061 ordered_multi_queue.cc:155] Queue waiting for data: (0, points2)
[playbag-6] process has finished cleanly
log file: /home/nitramek/.ros/log/554263dc-d60e-11e7-8959-000c29996d44/playbag-6*.log

```

I have ran validation tool on my bag and it gives me warning that some IMU data is out of order. I wrote simple python script to correct that (different imu topic careful, just /imu in the corrected bag) but it doesnt work either and gives me the same output. I know that the data in original bag has holes but I dont think that is the issue here. Perhaps there is some issue with the Velodyne data?

Bag file:
[https://drive.google.com/file/d/1-M4xlAmPHiYXM7hswUKTyxgsDeDZeYR6/view?usp=sharing](https://drive.google.com/file/d/1-M4xlAmPHiYXM7hswUKTyxgsDeDZeYR6/view?usp=sharing)

Its validation output:
```
E1130 21:29:06.174769 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.176908 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.177104 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.178323 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.179481 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.179555 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.180317 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.180701 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.180915 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.183269 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.183558 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.185173 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.185521 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.186200 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.186295 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.186556 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.189653 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.189870 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.190237 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.190553 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.190971 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.191226 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.191300 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.192682 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.192729 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.195588 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.195682 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.198061 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.198262 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.199861 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.202419 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.203385 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.204231 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.205227 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.206496 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.207087 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.207506 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.208130 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.208892 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.209120 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
E1130 21:29:06.209249 21438 rosbag_validate_main.cc:128] Sensor with frame_id "imu_link" jumps backwards in time. Make sure that the bag contains the data for each frame_id sorted by header.stamp, i.e. the order in which they were acquired from the sensor.
I1130 21:29:06.210237 21438 rosbag_validate_main.cc:155] Time delta histogram for consecutive messages on topic "/android/tango1/imu" (frame_id: "imu_link"):
Count: 951  Min: -0.001000  Max: 0.022000  Mean: 0.010283
[-0.001000, 0.001300)	                ####	Count: 186 (19.558359%)	Total: 186 (19.558359%)
[0.001300, 0.003600)	                    	Count: 0 (0.000000%)	Total: 186 (19.558359%)
[0.003600, 0.005900)	                    	Count: 0 (0.000000%)	Total: 186 (19.558359%)
[0.005900, 0.008200)	                    	Count: 23 (2.418507%)	Total: 209 (21.976866%)
[0.008200, 0.010500)	             #######	Count: 336 (35.331230%)	Total: 545 (57.308098%)
[0.010500, 0.012800)	                ####	Count: 195 (20.504732%)	Total: 740 (77.812828%)
[0.012800, 0.015100)	                    	Count: 5 (0.525762%)	Total: 745 (78.338593%)
[0.015100, 0.017400)	                    	Count: 0 (0.000000%)	Total: 745 (78.338593%)
[0.017400, 0.019700)	                   #	Count: 64 (6.729758%)	Total: 809 (85.068352%)
[0.019700, 0.022000]	                 ###	Count: 142 (14.931651%)	Total: 951 (100.000000%)
I1130 21:29:06.210389 21438 rosbag_validate_main.cc:155] Time delta histogram for consecutive messages on topic "/velodyne_points" (frame_id: "velodyne"):
Count: 7244  Min: 0.000091  Max: 0.290574  Mean: 0.001374
[0.000091, 0.029139)	####################	Count: 7243 (99.986198%)	Total: 7243 (99.986198%)
[0.029139, 0.058188)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.058188, 0.087236)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.087236, 0.116284)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.116284, 0.145333)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.145333, 0.174381)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.174381, 0.203429)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.203429, 0.232477)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.232477, 0.261526)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.261526, 0.290574
```
IMU sort tool (for completeness)
[https://gist.github.com/nitramek/0841f003899f4f67ff2e1aa117c0ae9b](https://gist.github.com/nitramek/0841f003899f4f67ff2e1aa117c0ae9b)
(the timing in the "fixed" bag seems different when I look on it via rqt_bag, so I am not sure about this "tool")

"Fixed imu order" bag.
https://drive.google.com/file/d/1bF6GwM74tkdes_0lqZq_0KRYhluKsdu9/view?usp=sharing

"Fixed imu order" bag validation output
```
I1130 21:27:53.164693 21421 rosbag_validate_main.cc:155] Time delta histogram for consecutive messages on topic "/imu" (frame_id: "imu_link"):
Count: 951  Min: 0.000000  Max: 0.022000  Mean: 0.010283
[0.000000, 0.002200)	                ####	Count: 186 (19.558359%)	Total: 186 (19.558359%)
[0.002200, 0.004400)	                    	Count: 0 (0.000000%)	Total: 186 (19.558359%)
[0.004400, 0.006600)	                    	Count: 0 (0.000000%)	Total: 186 (19.558359%)
[0.006600, 0.008800)	                    	Count: 23 (2.418507%)	Total: 209 (21.976866%)
[0.008800, 0.011000)	             #######	Count: 348 (36.593060%)	Total: 557 (58.569927%)
[0.011000, 0.013200)	                ####	Count: 188 (19.768665%)	Total: 745 (78.338593%)
[0.013200, 0.015400)	                    	Count: 0 (0.000000%)	Total: 745 (78.338593%)
[0.015400, 0.017600)	                    	Count: 1 (0.105152%)	Total: 746 (78.443741%)
[0.017600, 0.019800)	                  ##	Count: 96 (10.094637%)	Total: 842 (88.538383%)
[0.019800, 0.022000]	                  ##	Count: 109 (11.461619%)	Total: 951 (100.000000%)
I1130 21:27:53.165081 21421 rosbag_validate_main.cc:155] Time delta histogram for consecutive messages on topic "/velodyne_points" (frame_id: "velodyne"):
Count: 7244  Min: 0.000091  Max: 0.290574  Mean: 0.001374
[0.000091, 0.029139)	####################	Count: 7243 (99.986198%)	Total: 7243 (99.986198%)
[0.029139, 0.058188)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.058188, 0.087236)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.087236, 0.116284)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.116284, 0.145333)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.145333, 0.174381)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.174381, 0.203429)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.203429, 0.232477)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.232477, 0.261526)	                    	Count: 0 (0.000000%)	Total: 7243 (99.986198%)
[0.261526, 0.2905
```

Last thing there is no movement of the test cart with VLP sensor in this bag too, it is just for example purposes right now.

**Comment by gaschler** [comment]
Hi,
first of all, thanks for providing a very well documented issue.

I downloaded cart_test_fixed.bag and inspected it with `rqt_bag ~/Downloads/cart_test_fixed.bag`.
The two sensors have header time stamps that differ by 100000 seconds.
(Look at `header.stamp.secs`)
Velodyne points are so far behind they are never dispatched.
Cartographer requires good synchronization (<20ms) between points and IMU.
This must be fixed.
Perhaps using a separate device (phone) as an IMU is not good enough for even a first test, not sure if you will be able to synchronize it well enough.

Also, there are gaps larger than 0.1 s in IMU and 0.5 s in points data, which is bad.
The topic names and IMU data look fine to me.

**Comment by nitramek** [comment]
Hello Gashchler, thanks for quick feedback! I only glanced at those stamps didnt see it, will check/fix the bag and report back.

**Comment by nitramek** [comment]
UPDATE:
I am using this android implementation https://github.com/rpng/android_sensors_driver and I have experimented with [Line 230 IMU publisher](https://github.com/rpng/android_sensors_driver/blob/master/sensors_driver/src/udel/rpng/sensors_driver/publishers/ImuPublisher.java#L230).

When I assigned the time to be just `System.currentTimeInMillis()`, I was able to see output through Cartographer. 

Thanks again.

I still have question regarding the IMU and point cloud, how fast the data rates should be? I have looked at the example 3D muzeum bag and the IMU is quite bit slower the PointCloud2s. Also what are you using to synchronize the IMU and PointCloud2 messages or where I can read about it?
